models/spring-ai-bedrock-converse/src/main/java/org/springframework/ai/bedrock/converse/BedrockProxyChatModel.java

```diff
@@ -0,0 +1,710 @@
+
+		ConverseRequest converseRequest = this.createRequest(prompt);
+
+		ChatModelObservationContext observationContext = ChatModelObservationContext.builder()
+			.prompt(prompt)
+			.provider(AiProvider.BEDROCK_CONVERSE.value())
+			.requestOptions(buildRequestOptions(converseRequest))
+			.build();
+
+		ChatResponse chatResponse = ChatModelObservationDocumentation.CHAT_MODEL_OPERATION
+			.observation(this.observationConvention, DEFAULT_OBSERVATION_CONVENTION, () -> observationContext,
+					this.observationRegistry)
+			.observe(() -> {
+
+				ConverseResponse converseResponse = this.bedrockRuntimeClient.converse(converseRequest);
+
+				var response = this.toChatResponse(converseResponse);
+
+				observationContext.setResponse(response);
+
+				return response;
+			});
+
+		if (!this.isProxyToolCalls(prompt, this.defaultOptions) && chatResponse != null
+				&& this.isToolCall(chatResponse, Set.of("tool_use"))) {
+			var toolCallConversation = this.handleToolCalls(prompt, chatResponse);
+			return this.call(new Prompt(toolCallConversation, prompt.getOptions()));
+		}
+
+		return chatResponse;
+	}
+
+	private ChatOptions buildRequestOptions(ConverseRequest request) {
+		return ChatOptionsBuilder.builder()
+			.withModel(request.modelId())
+			.withMaxTokens(request.inferenceConfig().maxTokens())
+			.withStopSequences(request.inferenceConfig().stopSequences())
+			.withTemperature(request.inferenceConfig().temperature() != null
+					? request.inferenceConfig().temperature().doubleValue() : null)
+			.withTopP(request.inferenceConfig().topP() != null ? request.inferenceConfig().topP().doubleValue() : null)
+			.build();
+	}
+
+	@Override
+	public ChatOptions getDefaultOptions() {
+		return this.defaultOptions;
+	}
+
+	public ConverseStreamRequest createStreamRequest(Prompt prompt) {
+
+		ConverseRequest converseRequest = this.createRequest(prompt);
+
+		return ConverseStreamRequest.builder()
+			.modelId(converseRequest.modelId())
+			.messages(converseRequest.messages())
+			.system(converseRequest.system())
+			.additionalModelRequestFields(converseRequest.additionalModelRequestFields())
+			.toolConfig(converseRequest.toolConfig())
+			.build();
+	}
+
+	ConverseRequest createRequest(Prompt prompt) {
+
+		Set<String> functionsForThisRequest = new HashSet<>();
+
+		List<Message> instructionMessages = prompt.getInstructions()
+			.stream()
+			.filter(message -> message.getMessageType() != MessageType.SYSTEM)
+			.map(message -> {
+				if (message.getMessageType() == MessageType.USER) {
+					List<ContentBlock> contents = new ArrayList<>();
+					if (message instanceof UserMessage) {
+						var userMessage = (UserMessage) message;
+						contents.add(ContentBlock.fromText(userMessage.getContent()));
+
+						if (!CollectionUtils.isEmpty(userMessage.getMedia())) {
+							List<ContentBlock> mediaContent = userMessage.getMedia().stream().map(media -> {
+								ContentBlock cb = ContentBlock.fromImage(ImageBlock.builder()
```

Comment:
Bedrock also supports documents, it would be nice to have this feature implemented too. In my use-case I solved it with a convertMediaToContentBlock method:
```
        if (IMAGE.equals(mimeType.getType())) {
            return ContentBlock.fromImage(image -> image.format(
                            media.getMimeType().getSubtype())
                    .source(ImageSource.fromBytes(SdkBytes.fromByteArray(getContentMediaData(media.getData())))));
        } else {
            String fileName = mimeType.getParameter(FILE_NAME_KEY);
            return ContentBlock.fromDocument(document -> document.format(
                            FileNameUtils.detectDocFormatByFileName(fileName))
                    .name(FilenameUtils.getBaseName(fileName))
                    .source(DocumentSource.fromBytes(SdkBytes.fromByteArray(getContentMediaData(media.getData())))));
        }
```
possibly passing the filename in the mimeType is not general enough